File: 2026POTD/FEBRUARY/02-02-2026/3013.py
class Solution:
    def minimumCost(self, nums: List[int], k: int, dist: int) -> int:
        ans = float("inf")
        n = len(nums)
        i = 2
        j = 2 + dist
        total = 0
        minHeap = []
        maxHeap = []
        s = set()
        for p in range(i, j):
            heapq.heappush(minHeap, (nums[p], p))
        for _ in range(k - 2):
            val, idx = heapq.heappop(minHeap)
            total += val
            s.add(idx)
            heapq.heappush(maxHeap, (-val, idx))
        while n - i > k - 2:
            ans = min(ans, total + nums[0] + nums[i - 1])
            while minHeap and minHeap[0][1] <= i:
                heapq.heappop(minHeap)
            while maxHeap and maxHeap[0][1] <= i:
                heapq.heappop(maxHeap)
            if j < n:
                if maxHeap and nums[j] < -maxHeap[0][0]:
                    val, idx = heapq.heappop(maxHeap)
                    val = -val
                    heapq.heappush(minHeap, (val, idx))
                    s.remove(idx)
                    total -= val
                    total += nums[j]
                    heapq.heappush(maxHeap, (-nums[j], j))
                    s.add(j)
                else:
                    heapq.heappush(minHeap, (nums[j], j))
                j += 1
            if i in s:
                s.remove(i)
                total -= nums[i]
                val, idx = heapq.heappop(minHeap)
                total += val
                s.add(idx)
                heapq.heappush(maxHeap, (-val, idx))
            i += 1
        ans = min(ans, total + nums[0] + nums[i - 1])
        return ans

File: 2026POTD/FEBRUARY/02-02-2026/3013.cpp
#include <vector>
#include <set>
#include <numeric>
#include <algorithm>

using namespace std;

class Solution {
    long long current_sum = 0;
    multiset<int> left_set, right_set;
    int max_size;

    void balance() {
        while (left_set.size() < max_size && !right_set.empty()) {
            int val = *right_set.begin();
            current_sum += val;
            left_set.insert(val);
            right_set.erase(right_set.begin());
        }
        while (left_set.size() > max_size) {
            int val = *left_set.rbegin();
            current_sum -= val;
            right_set.insert(val);
            left_set.erase(prev(left_set.end()));
        }
    }

    void add(int val) {
        if (!left_set.empty() && val < *left_set.rbegin()) {
            current_sum += val;
            left_set.insert(val);
        } else {
            right_set.insert(val);
        }
        balance();
    }

    void remove(int val) {
        auto it = left_set.find(val);
        if (it != left_set.end()) {
            current_sum -= val;
            left_set.erase(it);
        } else {
            right_set.erase(right_set.find(val));
        }
        balance();
    }

public:
    long long minimumCost(vector<int>& nums, int k, int dist) {
        int n = nums.size();
        max_size = k - 2;
        current_sum = 0;
        left_set.clear();
        right_set.clear();

        for (int i = 2; i <= dist + 1; ++i) {
            add(nums[i]);
        }

        long long min_cost = nums[0] + nums[1] + current_sum;

        for (int i = 2; i <= n - (k - 1); ++i) {
            remove(nums[i]);
            if (i + dist < n) {
                add(nums[i + dist]);
            }
            min_cost = min(min_cost, (long long)nums[0] + nums[i] + current_sum);
        }

        return min_cost;
    }
};